ray_on_golem/server/services/ray.py

```diff
@@ -216,6 +228,35 @@ def _get_new_node_id(self) -> NodeId:
         self._nodes_id_counter += 1
         return node_id
 
+    async def _monitor_node(self, node_id: NodeId, activity: Activity) -> None:
+        async with self._get_node_context(node_id) as node:  # type: Node
+            is_head_node = self._is_head_node(node.tags)
+        while True:
+            try:
+                activity_state = await activity.get_state()
+                if (
+                    "Terminated" in activity_state.state
+                    or "Unresponsive" in activity_state.state
+                    or activity_state.error_message is not None
+                ):
+                    raise RayServiceError(
+                        f"Something is wrong with node {node_id} activity. {activity_state=}"
+                    )
+            except Exception:
+                msg = (
+                    f"Activity on {await self._golem_service.get_provider_desc(activity)} "
+                    f"is no longer accessible. Terminating {node_id} {is_head_node=}"
+                )
+                logger.warning(msg)
+                logger.debug(msg, exc_info=True)
+                del self._node_monitoring_tasks[node_id]
+                if is_head_node:
+                    await self.shutdown()
```

Comment:
This probably needs `async with self._nodes_lock:` when I think about it. @approxit is there a better way to terminate entire cluster?